rust/src/fees.rs

```diff
@@ -0,0 +1,216 @@
+const LABEL_SIZE: usize = 1;
+const CBOR_TAG: usize = 2;
+const ADDRESS: usize = 2 * HASH_OBJ;
+const CREDENTIAL: usize = LABEL_SIZE + HASH_OBJ;
+const UNIT_INTERVAL: usize = CBOR_TAG + SMALL_ARRAY + UINT + UINT;
+const FEE_SIZE: usize = LABEL_SIZE + UINT;
+const TTL_SIZE: usize = LABEL_SIZE + UINT;
+
+fn multisig_node_count(root: &Script) -> usize {
+    match &root.0 {
+        ScriptEnum::ScriptKeyNode(_) => 1,
+        ScriptEnum::ScriptAllOfNode(ScriptAllOfNode { scripts }) => 1usize + scripts.0.iter().map(multisig_node_count).sum::<usize>(),
+        _ => 2,
+    }
+}
+
+fn cert_size(cert: &DelegationCertificate) -> usize {
+    let x = match &cert.0 {
+        DelegationCertificateEnum::StakeKeyReg(_) |
+        DelegationCertificateEnum::StakeScriptKeyReg(_) => {
+            println!("wtf {} + {} + {} = {}", SMALL_ARRAY, LABEL_SIZE, HASH_OBJ, SMALL_ARRAY + LABEL_SIZE + HASH_OBJ);
+            SMALL_ARRAY + LABEL_SIZE + HASH_OBJ
+        },
+        DelegationCertificateEnum::StakeKeyDereg(_) |
+        DelegationCertificateEnum::StakeScriptKeyDereg(_) => SMALL_ARRAY + LABEL_SIZE + HASH_OBJ,
+        DelegationCertificateEnum::StakeDeleg(_) |
+        DelegationCertificateEnum::StakeScriptDeleg(_) => SMALL_ARRAY + LABEL_SIZE + 2 * HASH_OBJ,
+        DelegationCertificateEnum::PoolRegistration(PoolRegistration{ pool_params, .. }) => {
+            SMALL_ARRAY + LABEL_SIZE + HASH_OBJ
+            // pool owners
+            + CBOR_TAG + ARRAY_PREFIX + pool_params.owners.len() * HASH_OBJ
+            // cost
+            + UINT
+            // margin
+            + UNIT_INTERVAL
+            // pledge
+            + UINT
+            // operator
+            + HASH_OBJ
+            // vrf keyhash
+            + HASH_OBJ
+            // reward acc
+            + CREDENTIAL
+            // relays (String::len() is in bytes already, no need for utf8 conversion)
+            + pool_params.relays.0.iter().map(|url| url.len()).sum::<usize>()
+            // metadata (String::len() is in bytes, and metadat should only have 1 element)
+            + SMALL_ARRAY + pool_params.metadata.0.first().map_or(0, |pmd| pmd.url.len())
+        },
+        DelegationCertificateEnum::PoolRetirement(_) => SMALL_ARRAY + LABEL_SIZE + UINT + HASH_OBJ,
+        DelegationCertificateEnum::GenesisKeyDeleg(_) => SMALL_ARRAY + 2 * LABEL_SIZE + HASH_OBJ,
+        DelegationCertificateEnum::MoveRewardsCert(mir) => {
+            SMALL_ARRAY + LABEL_SIZE + MAP_PREFIX + mir.move_instantaneous_reward.len() * (UINT + HASH_OBJ)
+        },
+    };
+    println!("cert_size({:?}) = {}", cert, x);
+    x
+}
+
+#[wasm_bindgen]
+pub fn txsize(tx: &TransactionBody, witness: &TransactionWitnessSet) -> usize {
+    // It seems that in Haskell we have Tx = {Body, vkeySigs, msigScripts, metadata}
+    // but in the CDDL we have Body as one then vkeySigs/msigScripts as another
+    // as the data is segwit.
+    // We might want to make our own version that wraps these then.
```

Comment:
is this required? What is the downside of not doing this?

Reply:
I don't think there's much of a downside. It'd be similar to that `Transaction` or signed transaction from js-chain-libs. We were just in a very early stage that also had changing CDDL specs so I never implemented library usability things like this.